Add word_stem Presto function

yhwang opened the pull request:
Add snowball libstemmer v2.2.0 as one of the dependencies.
And use it to implement the word_stem() as a scalar UDF.

When using the libstemmer API, each language creates an `sb_stemmer`
instance which consumes 114 bytes, including the default 10 bytes for the output stem.
It uses the `realloc` to increase the memory block for the output stem if needed.

Fixes #8487

yhwang commented:
/cc @mbasmanova   @aditi-pandit

yhwang commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+  /// or create a new one if it doesn't exist. Return NULL if the
+  /// specified lang is not supported.
+  static Stemmer* getStemmer(const char* lang) {
+    thread_local std::map<std::string, Stemmer*> stemmers;
```
I don't clean up this thread-local map which contains the Stemmer instances for each language. I assume a working thread is shutting down with the Presto process. Not sure if it's worth adding the cleanup code while tearing down a thread.

mbasmanova replied:
I'm not sure if it is a good idea to use thread local variable here. CC: @bikramSingh91

mbasmanova replied:
https://github.com/zvelo/libstemmer

```
Creating a stemmer is a relatively expensive operation - the expected
usage pattern is that a new stemmer is created when needed, used
to stem many words, and deleted after some time.

Stemmers are re-entrant, but not threadsafe.  In other words, if
you wish to access the same stemmer object from multiple threads,
you must ensure that all access is protected by a mutex or similar
device.
```

Based on this information, it seems we could add a member variable to 'WordStemFunction' to store Stemmer instances. 

Let's use std::unique_ptr to make cleanup automatic.

Similar to regex functions, we may consider only supporting constant  values for 'lang' and create a single Stemmer instance in 'initialize'.

Alternatively, we may allow up to <small number> of different languages per function instance. See https://facebookincubator.github.io/velox/functions/presto/regexp.html

yhwang replied:
> Based on this information, it seems we could add a member variable to 'WordStemFunction' to store Stemmer instances.

that's the reason I use thread-local to store the lang<-->stemmer map

> Let's use std::unique_ptr to make cleanup automatic.

let me change the map to `<string, unique_ptr<Stemmer>>`

bikramSingh91 replied:
To add more context against using thread_local variables liberally is that we observed that storing thread local variables inside objects that can be moved between threads can cause an unexpected increase in memory usage as whenever they are accessed in another thread, they would be created and would exist for the lifetime of the thread. eg of an issue we recently fixed: https://github.com/facebookincubator/velox/pull/7646

In this case as well, the driver executing expression eval (and this code) can bounce between multiple available threads (we use folly::executor to execute drivers), and each time it would end up creating new instances. Unlike #7646, I do not expect this hold on to quite as much memory but nevertheless it would be a good idea to avoid this kind of usage pattern unless there is no other alternative. Using a unique_ptr as @mbasmanova suggested would help us to have tighter control over its lifetime by tying it to the lifetime of the Expression and therefore more predictable memory usage.

yhwang commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+    std::string lowerOutput;
+    stringImpl::lower<isAscii>(lowerOutput, input);
+    auto rev = stemmer->stem(lowerOutput);
+    if (rev == NULL) {
```
I guess I could put `UNLIKELY` here.

mbasmanova reviewed (commented):
@kgpai @assignUser Krishna, Jacob, would you help review build changes?

@yhwang let's add the new dependency to CMake/resolve_dependency_modules/README.md

@kgpai Krishna, would you help look into what would it take to add this dependency to Meta's repo?

mbasmanova commented on velox/docs/functions/presto/string.rst:
```diff
@@ -255,6 +255,14 @@ String Functions
 
+.. function:: word_stem(word) -> varchar
+
+    Returns the stem of ``word`` in the English language.
```
What happens if 'word' is not an English word?

yhwang replied:
it returns the same value

mbasmanova replied:
Please, update the documentation to clarify.

yhwang replied:
fixed in the new commit

mbasmanova commented on velox/docs/functions/presto/string.rst:
```diff
@@ -255,6 +255,14 @@ String Functions
+
+.. function:: word_stem(word, lang) -> varchar
+
+    Returns the stem of ``word`` in the ``lang`` language.
```
What are valid values for 'lang'? What happens if 'lang' is not valid?

yhwang replied:
throws an invalid arg exception

mbasmanova replied:
Please, update the documentation to clarify.

Also, what are valid values for 'lang'?

yhwang replied:
sure. the content is from Presto Java: http://prestodb.io/docs/0.285.1/functions/string.html. no valid values for lang there. but I certainly can add that in here.

mbasmanova replied:
Presto docs are often lacking important details. Let's make sure Velox docs are more complete.

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -15,6 +15,8 @@
  */
 #pragma once
 
+#include <libstemmer.h>
```
Let's create a separate file for this function.

yhwang replied:
create a new file named: `WordStem.h`

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
 
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
+/// to get word stem
+class Stemmer {
```
Let's put this into "details" namespace to make it clear that this is an implementation details and not a public API.

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+/// to get word stem
+class Stemmer {
+ private:
+  sb_stemmer* sbstemmer;
```
Please, review coding style and update to comply.

yhwang replied:
fixed

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+      stemmers[lang] = rev;
+      return rev;
+    }
+    return NULL;
```
nullptr, please

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+    return NULL;
+  }
+
+  /// Get the word stem or NULL if out of memory
```
What does "out of memory" mean here? Do we use Velox's MemoryPool to allocate memory?

yhwang replied:
the stemmer lib should use the native one unless velox overrides the `new` operator. I enclosed the API doc in another comment and `out of memory` is from its API doc.

mbasmanova replied:
> the stemmer lib should use the native one unless velox overrides the new operator. 

Velox doesn't override 'new'. Does stemmer library allow for specifying custom memory allocator?

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+  }
+
+  /// Get the word stem or NULL if out of memory
+  const char* stem(const std::string& input) {
```
Can we pass std::string_view or StringView instead of std::string to avoid unnecessary copy?

yhwang replied:
line 598 is using std::string, so I pass std::string& here.

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+/// It provides Java implementation which is used in Presto as well
+/// as C implementation. Therefore, both Presto and Prestimissio
+/// would have the same word stem results.
+template <typename T>
```
nit: T -> TExec

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+    auto stemmer = Stemmer::getStemmer(lang);
+    if (!stemmer) {
+      // language is not supported
+      throw std::invalid_argument(
```
use VELOX_USER_FAIL

yhwang replied:
good tip. let me also explore other macros for future use. thanks.

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+
+    std::string lowerOutput;
+    stringImpl::lower<isAscii>(lowerOutput, input);
+    auto rev = stemmer->stem(lowerOutput);
```
rev -> stem

please, review coding guidelines

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+    stringImpl::lower<isAscii>(lowerOutput, input);
+    auto rev = stemmer->stem(lowerOutput);
+    if (rev == NULL) {
+      throw std::runtime_error("out of memory");
```
Would you explain how memory allocations are happening here? We need to make sure any large allocations go through Velox's Memory Pool. How can we run out of memory here?

yhwang replied:
enclosed its API doc here:

```
/** Stem a word.
 *
 *  The return value is owned by the stemmer - it must not be freed or
 *  modified, and it will become invalid when the stemmer is called again,
 *  or if the stemmer is freed.
 *
 *  The length of the return value can be obtained using sb_stemmer_length().
 *
 *  If an out-of-memory error occurs, this will return NULL.
 */
const sb_symbol *   sb_stemmer_stem(struct sb_stemmer * stemmer,
                                    const sb_symbol * word, int size);
```

that's why I throw an `out of memory` exception when the returned value is NULL.

mbasmanova replied:
Let's use VELOX_CHECK_NOT_NULL(stem, "Stemmer library returned a NULL (out-of-memory)."

We need to make it clear that the error comes from Stemmer library and not from Velox core.

Let's also figure out how much memory Stemmer instances use.

yhwang replied:
> Let's use VELOX_CHECK_NOT_NULL(stem, "Stemmer library returned a NULL (out-of-memory)."

Sure thing!

> Let's also figure out how much memory Stemmer instances use.

I guess this would be a tricky one. I will dig more on this.

mbasmanova replied:
You can write a simple program that uses stemmer lib to stem some words and capture process-wide memory usage before creating stemmer instance, after, then again after processing some words, then after deleting stemmer instance.

https://stackoverflow.com/questions/63166/how-to-determine-cpu-and-memory-consumption-from-inside-a-process

yhwang replied:
I can also look into its implementation and see how and how much it allocates for storing the stem. since each stemmer reuses the memory block, it shouldn't consume that much.

mbasmanova commented on velox/functions/prestosql/tests/StringFunctionsTest.cpp:
```diff
@@ -305,6 +305,12 @@ class StringFunctionsTest : public FunctionBaseTest {
       const std::string& left,
       const std::string& right);
 
+  std::string wordStemWithLang(
```
Let's create a separate test file.

mbasmanova reviewed (commented):
@yhwang Thank you for working on this. Some initial questions. Primarily, I'd like to understand how we are allocating memory and what may cause OOM scenario.

CC: @xiaoxmeng

mbasmanova commented:
CC: @amitkdutta

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+struct WordStemFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
```
It doesn't look like we are reusing inputs for result. It looks like we are making a copy. Hence, let's remove this property.

yhwang replied:
sure

assignUser commented on CMake/resolve_dependency_modules/stemmer.cmake:
```diff
@@ -0,0 +1,49 @@
+add_library(stemmer STATIC IMPORTED)
+set_target_properties(stemmer PROPERTIES IMPORTED_LOCATION ${BINARY_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}stemmer${CMAKE_STATIC_LIBRARY_SUFFIX})
+
+include_directories(${BINARY_DIR}/include)
```
```suggestion
target_include_directories(stemmer PUBLIC ${BINARY_DIR}/include)
```
No global includes please

yhwang replied:
fixed

assignUser commented on CMake/resolve_dependency_modules/stemmer.cmake:
```diff
@@ -0,0 +1,49 @@
+
+ExternalProject_Get_Property(libstemmer BINARY_DIR)
+
+add_library(stemmer STATIC IMPORTED)
```
```suggestion
add_library(stemmer STATIC IMPORTED)
add_library(stemmer::stemmer ALIAS stemmer)
```
Please add a namespaced target and use that to link against.

assignUser commented on velox/functions/prestosql/CMakeLists.txt:
```diff
@@ -74,7 +74,8 @@ target_link_libraries(
   velox_functions_util
-  Folly::folly)
+  Folly::folly
+  stemmer)
```
```suggestion
  stemmer::stemmer)
```

assignUser commented on CMakeLists.txt:
```diff
@@ -542,6 +542,10 @@ endif()
 set_source(xsimd)
 resolve_dependency(xsimd 10.0.0)
 
+set(stemmer_SOURCE BUNDLED)
+set_source(stemmer)
```
```suggestion
set(stemmer_SOURCE BUNDLED)
```
Looking at the library source they don't seem to provide any form of config or pc file ... so that would require us to create a Findstemmer.cmake. Depending on compile time it might be worth it, if it's fast we can skip it though and always bundle for now.

yhwang replied:
yea, the build is pretty fast. So I leave it as-is.

assignUser commented on CMakeLists.txt:
```diff
@@ -542,6 +542,10 @@ endif()
 
+set(stemmer_SOURCE BUNDLED)
+set_source(stemmer)
+resolve_dependency(stemmer)
```
This is now a default addition that is required for all builds? Please also add it to CMake/resolve_dependency/README.md

yhwang replied:
done!

assignUser commented on CMake/resolve_dependency_modules/stemmer.cmake:
```diff
@@ -0,0 +1,49 @@
+  INSTALL_COMMAND ""
+  BUILD_BYPRODUCTS ${STEMMER_PREFIX}/src/libstemmer/${CMAKE_STATIC_LIBRARY_PREFIX}stemmer${CMAKE_STATIC_LIBRARY_SUFFIX})
+
+ExternalProject_Get_Property(libstemmer BINARY_DIR)
```
As you have set BUILD_IN_SOURCE that property should be empty but might be auto-populated as the build passes.

yhwang replied:
You're right. Since `BUILD_IN_SOURCE` is set, I set a property to point to the include dir instead.

assignUser reviewed (changes requested):
Some comments for the cmake. I checked the timestamps in the debug build and stemmer build in a few seconds, so we can  leave it on BUNDLED for now.

yhwang commented:
thanks for the comments. I will address them in separate commits for an easier review process. I will squash all commits into one when the code is ready for merge.

@assignUser I addressed your comments in this commit: [Address Cmake comments](https://github.com/facebookincubator/velox/pull/9363/commits/e927a8df2d2078f4a0e5312a6c5d9e33fff166e7)

I will work on other comments later.

mbasmanova commented:
@rui-mo @PHILO-HE Folks, does Spark have a similar function?

assignUser commented:
The python extension is always build as a shared library, could we build libstemmer with fpic? https://github.com/facebookincubator/velox/actions/runs/8626883438/job/23645998539?pr=9363#step:13:3508

yhwang commented:
> The python extension is always build as a shared library, could we build libstemmer with fpic? https://github.com/facebookincubator/velox/actions/runs/8626883438/job/23645998539?pr=9363#step:13:3508

@assignUser good question. I didn't think about that since libstemmer comes with the Makefile that only builds static lib. Do you think we should tweak it to build a so?

mbasmanova commented:
@yhwang 

> use local_thread or not

Definitely not. Please, take a look at @bikramSingh91's explanation.

yhwang commented:
> @yhwang
> 
> > use local_thread or not
> 
> Definitely not. Please, take a look at @bikramSingh91's explanation.

I guess my point is I don't see a direct relationship between this change and the thread-local issue mentioned above. I'd like to get more clarification. The stemmers in the use case here would only take a fixed amount of memory, it won't grow infinitely. Just need to know if that's the main concern.

assignUser commented:
@yhwang it doesn't look like it supports shared library but patching the make file with
```diff
- CFLAGS=-Iinclude
+ CFLAGS=-Iinclude -fPIC
```
should be enough for the linker to figure stuff out when building the extension .so

bikramSingh91 commented:
> > @yhwang
> > > use local_thread or not
> > 
> > 
> > Definitely not. Please, take a look at @bikramSingh91's explanation.
> 
> I guess my point is I don't see a direct relationship between this change and the thread-local issue mentioned above. I'd like to get more clarification. The stemmers in the use case here would only take a fixed amount of memory, it won't grow infinitely. Just need to know if that's the main concern.

It appears that a static method is responsible for maintaining the thread-local variables specific to each thread. These variables are accessed each time the simple function is invoked. Although these variables are not contained within an object, the driver that carries this expression can be executed on a different thread for the subsequent batch of input vectors, leading to the creation of additional instances.

The maximum memory these objects can hold is approximately equal to (num_of_langs * threads). While this might constitute a small amount, the presence of thread-local variables complicates the understanding of their lifespan, ownership, and the number of copies created. The alternative we proposed ties the ownership to the Expression object itself, thereby making these aspects more predictable while ensuring thread-safe access. TLDR:  the concern is less about the memory utilization and more about developing code that we can confidently reason about.

We would appreciate any insights into specific use-cases or advantages that we might have overlooked, which would make the use of thread-local variables more appealing. For instance, is their purpose to ensure that these objects are shared among potentially different expressions objects within the same thread to save on the cost of creating the stemmer objects?

yhwang commented:
@bikramSingh91 thanks for the clarification. let me simplify it to what the lifespan of a stemmer should be.
- with the expression:
  initialize stemmer(s) in each expression and clean them up after the expression ends
- or with the thread:
  lazy initialize stemmers when needed and use unique_ptr for it. Basically, they go down with the thread

For me, maintaining the logic in the WordStem UDF struct is more straightforward and clean. Not sure if expression usage makes the code separate in multiple places. But I am good with either choice.

 > We would appreciate any insights into specific use-cases or advantages that we might have overlooked, which would make the use of thread-local variables more appealing.

This is the API doc from libstemmer:
```
Creating a stemmer is a relatively expensive operation - the expected
usage pattern is that a new stemmer is created when needed, used
to stem many words, and deleted after some time.

Stemmers are re-entrant, but not threadsafe.  In other words, if
you wish to access the same stemmer object from multiple threads,
you must ensure that all access is protected by a mutex or similar
device.
```
@mbasmanova posted this above and this is also the main reason I use thread-local map to store the stemmers, avoid mutex, and keep the code simple to read. Again, I am good with either expression or thread-local, since we agree memory shouldn't be the concern.

yhwang commented:
> @yhwang it doesn't look like it supports shared library but patching the make file with
> 
> ```diff
> - CFLAGS=-Iinclude
> + CFLAGS=-Iinclude -fPIC
> ```
> 
> should be enough for the linker to figure stuff out when building the extension .so

@assignUser thanks for the idea of the `-fPIC` flag. But I found the Makefile is made for the static lib(it uses `ar`). To have a shared lib, an extra `gcc -shared` is needed. Or you can share some cmake tricks to achieve that :-p.

assignUser commented:
@yhwang afaiu the issue is not that it's a static library (all of velox is static libs after all and still part of the extension.so) but that it contains non-relocatable code and that should be fixed by -fPIC

PHILO-HE commented:
> @rui-mo @PHILO-HE Folks, does Spark have a similar function?

@mbasmanova, Spark has no such function. cc @rui-mo.

yhwang commented:
> @yhwang afaiu the issue is not that it's a static library (all of velox is static libs after all and still part of the extension.so) but that it contains non-relocatable code and that should be fixed by -fPIC

@assignUser I see your point now. do you suggest that I patch it to have `-fPIC`?

assignUser commented:
@yhwang Yes, you can check out some of the other cmake files in resovle_dependency for inspiration, it works the same for external project and fetch content.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,128 @@
+/// as C implementation. Therefore, both Presto and Prestimissio
+/// would have the same word stem results.
+template <typename TExec>
+struct WordStemFunction {
```
@yhwang All you need to do is the following:

1- Add member variable to WordStemFunction to store a map of stemmers.
2- Move getStemmer method into WordStemFunction struct as a private method.

yhwang replied:
When I move getStemmer into WordStemFunction, it moves the `std::map<std::string, std::unique_ptr<Stemmer>> stemmers` together. Then wait for the thread-local decision, right?

mbasmanova replied:
> Then wait for the thread-local decision, right?

Not sure if we need to wait. Let's remove 'thread-local' and make the map a regular private member of the WordStemFunction struct.

FYI, in case you haven't seen these already: https://facebookincubator.github.io/velox/develop/scalar-functions.html

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,128 @@
+    auto stemmer = Stemmer::getStemmer(lang);
+    if (!stemmer) {
+      // language is not supported
+      VELOX_USER_FAIL("Unknown stemmer language: \"{}\"", lang);
```
use auto* + VELOX_USER_NOT_NULL

```
auto* stemmer = Stemmer::getStemmer(lang);
VELOX_USER_NOT_NULL(stemmer, "Unsupported stemmer language: {}", lang)
```

mbasmanova replied:
Yup.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,128 @@
+
+    std::string lowerOutput;
+    stringImpl::lower<isAscii>(lowerOutput, input);
+    auto stem = stemmer->stem(lowerOutput);
```
nit: auto*

mbasmanova commented on velox/functions/prestosql/tests/WordStemTest.cpp:
```diff
@@ -0,0 +1,78 @@
+
+using namespace facebook::velox::functions::test;
+
+class WordStemTest : public FunctionBaseTest {
```
Put this class into anonymous namespace inside facebook::velox::functions

```
namespace facebook::velox::functions {
namespace {

class WordStemTest ...
}
}
```

mbasmanova commented on velox/functions/prestosql/tests/WordStemTest.cpp:
```diff
@@ -0,0 +1,78 @@
+
+class WordStemTest : public FunctionBaseTest {
+ protected:
+  std::string wordStemWithLang(
```
nit: drop "WithLang" suffix

yhwang commented:
@assignUser added `-fPIC` CPPFlags for libstemmer, please check this commit: [Add -fPIC CPPFLAGS](https://github.com/facebookincubator/velox/pull/9363/commits/74be54ef872fff94b7750beb037afc870386b45c)
Thanks.

Work on the next one :-)

assignUser commented:
Thanks looks good, there is currently an issue with the linux workflows so we can't test it in ci atm, once https://github.com/facebookincubator/velox/pull/9451 is merged please rebase/merge those :)

yhwang commented:
@mbasmanova addressed your comments on this commit: [Address WordStem comments](https://github.com/facebookincubator/velox/pull/9363/commits/b46be4dd858fdc283be90138382ecb5c2c026ddc)

I will update you on the memory allocation usage later.

yhwang commented:
@mbasmanova for the memory allocation:

libstemmer is using `malloc`, `calloc`, and `realloc` for the memory allocators. The default size of a stemmer is 114 bytes, including the default 10 bytes for the output stem. It uses the `realloc` to increase the block if needed.

yhwang commented:
@mbasmanova in order to get the precise size for sb_stemmer, I added instructions into the source code to get the details. I can double-check with the approach from that discussion thread too.

yhwang commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,128 @@
+  /// or create a new one if it doesn't exist. Return nullptr if the
+  /// specified lang is not supported.
+  static Stemmer* getStemmer(const char* lang) {
+    thread_local std::map<std::string, std::unique_ptr<Stemmer>> stemmers;
```
done! [Move stemmers as a static member var](https://github.com/facebookincubator/velox/pull/9363/commits/54c6d567775279b4cee487f8b1c23ceab3e59738)

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,130 @@
+  static constexpr bool is_default_ascii_behavior = true;
+
+ private:
+  thread_local static inline std::map<std::string, std::unique_ptr<Stemmer>>
```
Please, remove thread_local and static.

yhwang replied:
? but it's needed from a static method. without thread_local for now, it will shared by multiple threads without mutex.

mbasmanova replied:
The method shouldn't be static either.

mbasmanova replied:
Check out regex function's for an example. They have a similar cache of compiled regexes.

yhwang replied:
I see. so switch to expression lifespan for the stemmers.

yhwang replied:
one question,  different drivers have their own WordStemFunction instance, right? make sure I don't need mutex here.

mbasmanova replied:
That's right. Each logical thread of execution, driver, will create its own instance of the function. Each function will run single-threaded (it may run on different threads, but not concurrently).

yhwang commented:
@mbasmanova here is the commit that switch from thread-local to expression lifespan for the stemmer: [use expression lifespan for stemmers](https://github.com/facebookincubator/velox/pull/9363/commits/8374a6fece209211b3c8b2810bd82aed42c5ad2f)

I want to run `velox_functions_test` with `--gtest_repeat` to compare the CPU time before and after. However, I found that I couldn't measure the cache with those test cases, basically, each `evaluateOnce` creates a new WordStemFunction instance. So, I didn't make any comparison between `thread-local` and `expression`

mbasmanova commented:
@yhwang To evaluate performance, you need to create a benchmark and run it on sufficiently large number of rows using release build. See velox/benchmarks/basic/LikeBenchmark.cpp for reference.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+
+namespace facebook::velox::functions {
+
+namespace {
```
No anonymous namespaces in headers. Use "details" namespace.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+namespace facebook::velox::functions {
+
+namespace {
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
```
Use /// for public APIs only. Since this is not public API, use //.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
+/// to get word stem
+class Stemmer {
+ private:
```
In a class, public methods come first, followed by private methods and member variables.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+
+ public:
+  Stemmer(sb_stemmer* stemmer) : sbStemmer_(stemmer) {}
+  ~Stemmer() {
```
add empty line between methods

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+  sb_stemmer* sbStemmer_;
+
+ public:
+  Stemmer(sb_stemmer* stemmer) : sbStemmer_(stemmer) {}
```
add a safety check VELOX_CHECK_NOT_NULL(stemmer)

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+    sb_stemmer_delete(sbStemmer_);
+  }
+
+  /// Get the word stem or NULL if out of memory
```
Make sure all comments end in a period.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  /// Get the word stem or NULL if out of memory
+  const char* stem(const std::string& input) {
```
Just curious, does this function return zero-terminated string?

yhwang replied:
yes!

mbasmanova replied:
Got it. Since this is a bit unusual in modern C++, it would be nice to clarify that in the comments.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+/// word_stem(word, lang) -> varchar
+///     return the stem of the word in the specificed language
+///
+/// It uses the snowball stemmer library to calculate the stem.
```
Drop "It": Uses ....

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+ private:
```
same comment about order of methods and variables

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+  static constexpr bool is_default_ascii_behavior = true;
+
+ private:
+  std::map<std::string, std::unique_ptr<Stemmer>> stemmers_;
```
use folly::F14FastMap (or std::unordered_map)

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+ private:
+  std::map<std::string, std::unique_ptr<Stemmer>> stemmers_;
+
+  /// Get a Stemmer from the the map using the lang as the key or create a new
```
//

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+    if (auto found = stemmers_.find(lang); found != stemmers_.end()) {
+      return found->second.get();
+    }
+    Stemmer* stemmer = nullptr;
```
Move this variable inside 'if'.

```
if (...) {
   auto* stemmer = new Stemmer(sbStemmer);
   stemmers_[lang] = std::unique_ptr<Stemmer>(stemmer);
   return stemmer;
}

return nullptr;
```

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,129 @@
+    result = stem;
+  }
+};
+} // namespace facebook::velox::functions
```
add empty line at the end of the file

mbasmanova commented on velox/functions/prestosql/tests/WordStemTest.cpp:
```diff
@@ -0,0 +1,80 @@
+}
+
+} // namespace
+} // namespace facebook::velox::functions
```
empty line at the end of the file

mbasmanova reviewed (commented):
@yhwang Thank you for iterating. Looks great % remaining style nits.

Please, update PR description to clarify which version of the stemmer you are adding and how much memory you believe each instance is using.

mbasmanova commented on CMake/resolve_dependency_modules/README.md:
```diff
@@ -39,6 +39,7 @@ by Velox. See details on bundling below.
 | fbthrift          | v2024.02.26.00  | No       |
 | DuckDB (testing)  | 0.8.1           | Yes      |
 | cpr (testing)     | 1.10.15         | Yes      |
+| libstemmer        | 2.2.0           | Yes      |
```
Should we move this before "testing" dependencies? CC: @kgpai @assignUser

assignUser commented:
Hm seems like our centos images are missing `patch` can you add a step prior to the build with `dnf install -y patch` and also add patch to the build_requirements in the `setup-*.sh` scripts? (this will add it to the docker container on merge and we can remove the additional step afterwards)

yhwang commented:
@mbasmanova  fixed code style in this: [fix code style](https://github.com/facebookincubator/velox/pull/9363/commits/b246add1bca85b2aeec119be74b163f084bda434)

I will update the commit description when I squash the commits.

@assignUser use `git apply` in this: [apply patch with git apply](https://github.com/facebookincubator/velox/pull/9363/commits/8e76469ed31ab3c4a31e93ba4f26b274efa4991f)

yes, `git apply` works, at least on my machine :-p

mbasmanova commented:
@yhwang 

> I will update the commit description when I squash the commits.

Velox repo works differently from PrestoDB repo. When merging a PR, a single commit is created using PR title and description to form a commit message. Hence, it is important to ensure the PR title and description are written well and work as a commit message.

I know that @assignUser likes separate commits for addressing comments, but I don't share that preference. When number of commits goes above 10-ish GitHub becomes very slow, hence, I suggest to either keep one commit or a small number of these.

mbasmanova commented on velox/functions/prestosql/WordStem.h:
```diff
@@ -0,0 +1,132 @@
+    sb_stemmer_delete(sbStemmer_);
+  }
+
+  // Get the word stem or NULL if out of memory.
```
> if out of memory

This sounds very strange given that you are saying stemmer uses < 200 bytes total. I wonder if this is just very old code from the days when 200 bytes was a lot of memory. However, this would be confusing to Velox developers today and trigger questions about why we don't actively manage memory allocations from stemmer.

I suggest to rephrase to avoid that confusion.

```
Returns the work stem or nullptr if an unlikely out-of-memory error occurs.
```

yhwang replied:
fixed

mbasmanova reviewed (approved):
@yhwang Looks good to me. Thank you for contributing this function and iterating on the PR.

yhwang commented:
@mbasmanova 

> Velox repo works differently from PrestoDB repo. 

Understood. Different repos usually have different rule/bot. I worked on other repos having bots to squash commits and use the PR description and commit messages as the final commit. Good to know rule/settings here.


> I know that @assignUser likes separate commits for addressing comments, but I don't share that preference. 

I guess there is a way to fulfill both, I could squash reviewed commits into one and have only new commits for review :-) . I use single commit from time to time. But I found the force push diff is not that user-friendly. If the change is small, I prefer a single commit too.

Thanks for the review and comments (@assignUser too). I learned a lot. It's also a good practice for me. I updated the comment and squashed my commits (I know I don't need to do that)

pedroerp commented:
@yhwang looks like our internal fuzzer caught a crash with your function:

```
I0418 07:57:34.180382 1011044 ExpressionFuzzerVerifier.cpp:312] ==============================> Started iteration 101 (seed: 2144156392)                                                       
I0418 07:57:34.189090 1011044 ExpressionVerifier.cpp:51] Executing expression 0 : word_stem("c0",MJw$$A,*3z;D)                                                                                 
=================================================================                                                                                                                              
==1011044==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7f258f990910 at pc 0x00000032eb39 bp 0x7fffc1c98860 sp 0x7fffc1c98020                                                   
READ of size 13 at 0x7f258f990910 thread T0                                                                                                                                                    
SCARINESS: 41 (multi-byte-read-stack-buffer-overflow)                                                                                                                                          
    #0 0x32eb38 in __interceptor_strlen.part.0 ubsan.c                                                                                                                                         
    #1 0x7f25ccb86507 in unsigned long folly::detail::constexpr_strlen_internal<char, 0ul>(char const*, int) fbcode/folly/portability/Constexpr.h:50                                           
    #2 0x7f25ccb864e6 in unsigned long folly::constexpr_strlen<char>(char const*) fbcode/folly/portability/Constexpr.h:106                                                                     
    #3 0x7f25ccb85209 in folly::Range<char const*>::Range(char const*) fbcode/folly/Range.h:303                                                                                                
    #4 0x7f25d6c69e62 in unsigned long folly::detail::TransparentRangeHash<char>::operator()<char const*>(char const* const&) const fbcode/folly/container/HeterogeneousAccess.h:125           
    #5 0x7f25d6c69a94 in unsigned long folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::unique_ptr<facebook::velo
x::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer>>, void, void, void, std::integral_constant<bool, true>>::computeKeyHash<char const*>(char const
* const&) const fbcode/folly/container/detail/F14Policy.h:1135                                                                                                                                 
    #6 0x7f25d6c678bb in find<const char *> fbcode/folly/container/detail/F14Table.h:1585                                                                                                      
    #7 0x7f25d6c678bb in find<const char *> fbcode/folly/container/F14Map.h:886                                                                                                                
    #8 0x7f25d6c678bb in facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::getStemmer(char const*) fbcode/velox/functions/prestosql/WordStem.h:120              
    #9 0x7f25d6d1b51f in doCall<true> fbcode/velox/functions/prestosql/WordStem.h:101                                                                                                          
    #10 0x7f25d6d1b51f in callAscii fbcode/velox/functions/prestosql/WordStem.h:93                                                                                                             
    #11 0x7f25d6d1b51f in callAsciiImpl fbcode/velox/core/SimpleFunctionMetadata.h:954                                                                                                         
    #12 0x7f25d6d1b51f in callAscii fbcode/velox/core/SimpleFunctionMetadata.h:904
    #13 0x7f25d6d1b51f in doApplyAsciiNotNull<2UL, facebook::velox::StringView, facebook::velox::StringView, 0> fbcode/velox/expression/SimpleFunctionAdapter.h:866
    #14 0x7f25d6d1b51f in doApplyAsciiNotNull<1UL, const facebook::velox::exec::ConstantVectorReader<facebook::velox::Varchar>, facebook::velox::StringView, 0> fbcode/velox/expression/SimpleF
unctionAdapter.h:855
    #15 0x7f25d6d1b51f in doApplyAsciiNotNull<0UL, facebook::velox::exec::FlatVectorReader<facebook::velox::Varchar>, facebook::velox::exec::ConstantVectorReader<facebook::velox::Varchar>, 0>
 fbcode/velox/expression/SimpleFunctionAdapter.h:855
    #16 0x7f25d6d1b51f in operator()<facebook::velox::exec::StringWriter<false>, int> fbcode/velox/expression/SimpleFunctionAdapter.h:685
    #17 0x7f25d6d1b51f in operator()<int> fbcode/velox/expression/SimpleFunctionAdapter.h:721
    #18 0x7f25d6d1b51f in operator()<int> fbcode/velox/expression/EvalCtx.h:102
```

You can reproduce it by running fuzzer with ubsan enabled.

yhwang commented:
@pedroerp thanks for the info. the code in `velox/functions/prestosql/WordStem.h:120` is:

```
if (auto found = stemmers_.find(lang); found != stemmers_.end()) {
```

It's kind of interesting to me. Let me reproduce it and see how to fix it.

pedroerp commented:
> Is the second parameter a string?

I suppose so. Try to reproduce by running on that exact same version and same fuzzer seed. You'll probably need to run using clang and enable ubsan.

yhwang commented:
@pedroerp sorry for the late response, I was on vacation last Friday. I tried to build `velox_expression_fuzzer_test` with UBSan today, but I hit some undefined symbol errors in the linking stage. Can you share the steps/instructions to build the fuzzer with UBSan enabled?

pedroerp commented:
@yhwang unfortunately we don't have this set up in our github CI (yet), but you need to first compile using clang with the ubsan flags enabled, then run join fuzzer. These are the instructions on how to enable ubsan:

https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html


What is the issue you are seeing?

yhwang commented:
@pedroerp 

> https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

Thanks and this is the same doc that I followed. The unresolved symbols issue may be caused by the flex that comes with xcode. I am using a Linux VM to build the velox now. The build looks good so far.

pedroerp commented:
> The unresolved symbols issue may be caused by the flex that comes with xcode

I think you need to compile all dependencies with the same ubsan flags. @kgpai and @assignUser would know this better.

yhwang commented:
@pedroerp 

I ran the following command:
`_build/debug/velox/expression/tests/velox_expression_fuzzer_test --seed=2144156392 --only=word_stem --logtostdout=1 --steps=1000`

and I only saw this from time to time:
```
I20240423 16:/home/yhwang/WS_git/presto/presto-native-execution/velox/velox/common/base/BitUtil.cpp:204:24: runtime error: load of misaligned address 0x7ffe3eb09834 for type 'const long unsigned int', which requires 8 byte alignment
0x7ffe3eb09834: note: pointer points here
  0a 00 00 00 59 28 46 76  27 77 30 3d 4a 66 00 00  00 00 00 00 00 00 00 00  87 98 b0 3e fe 7f 00 00
              ^~
```

and although I specified `-seed`, it only uses it for the first iteration. Do you know how to use the same seed across all iterations?

I added this to the very top CMakeLists.xt:
```
add_link_options(-fsanitize=undefined)
add_compile_options(-fsanitize=undefined)
```
And set `VELOX_DEPENDENCY_SOURCE` to `BUNDLED`

I guess this way it builds most of the deps as bundles and using the UBSan flag.

pedroerp commented:
> Do you know how to use the same seed across all iterations?

The seed is used for the first iteration, then a new seed is generated for the next iteration. They should deterministically reproduce the same chain of seeds if you're using the same binary with the same set of functions available.

yhwang commented:
Then I can only for-loop the command to see if I can hit that error with the same seed. So far, I haven't seen it yet.

kgpai commented:
@yhwang You can see the flags we use here : https://github.com/facebookincubator/velox/blob/main/.github/workflows/scheduled.yml#L360 


You can also set the flag in helper-functions.sh without modifying your cmake and set the deps to BUNDLED which will ensure everything has UBSAN (like you did!).

yhwang commented:
I applied all the flags, but I only hit this error still:
```
/home/yhwang/WS_git/presto/presto-native-execution/velox/./velox/common/base/BitUtil.h:826:12: runtime error: store to misaligned address 0x564e2ecaefc3 for type 'short unsigned int', which requires 2 byte alignment
0x564e2ecaefc3: note: pointer points here
 ff  fd ff e7 ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff
              ^
```

Based on the error that Pedro posted earlier, I guess it's caused by the `const char*` retrieving from StringViewer::data(). Let me see if passing std::string would avoid the error.

yhwang commented:
@pedroerp I kept trying to recreate the error you posted above. But I can't reproduce it. The only error I saw is the one I mentioned: `BitUtil.h:826`. So the only thing I can do is change the `lang` param from `const char*` to `std::string&`. Could you help me to verify the change at your end?

Another thing I feel weird is the stack trace in your comment is different from what I saw in my system. I paused at the word_stem function and stepped into the function call one by one to simulate the stack trace you posted. and this is what I got (I didn't go all the way to the very top stack trace you had):
```
folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> >::computeKeyHash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(const folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> > * const this, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > & key) (/home/yhwang/WS_git/presto/presto-native-execution/velox/_build/debug/_deps/folly-src/folly/container/detail/F14Policy.h:1134)
folly::f14::detail::F14Table<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> > >::find<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > & key, const folly::f14::detail::F14Table<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> > > * const this) (/home/yhwang/WS_git/presto/presto-native-execution/velox/_build/debug/_deps/folly-src/folly/container/detail/F14Table.h:1580)
folly::f14::detail::F14BasicMap<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> > >::find(const folly::f14::detail::F14BasicMap<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> > >::key_type & key, folly::f14::detail::F14BasicMap<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<facebook::velox::functions::detail::Stemmer, std::default_delete<facebook::velox::functions::detail::Stemmer> >, void, void, void, std::integral_constant<bool, true> > > * const this) (/home/yhwang/WS_git/presto/presto-native-execution/velox/_build/debug/_deps/folly-src/folly/container/F14Map.h:854)
facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::getStemmer(facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec> * const this, const std::string & lang) (/home/yhwang/WS_git/presto/presto-native-execution/velox/velox/functions/prestosql/WordStem.h:120)
facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::doCall<true>(const std::string & lang, const facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::arg_type & input, facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::out_type & result, facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec> * const this) (/home/yhwang/WS_git/presto/presto-native-execution/velox/velox/functions/prestosql/WordStem.h:101)
facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::callAscii(const facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::arg_type & input, facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec>::out_type & result, facebook::velox::functions::WordStemFunction<facebook::velox::exec::VectorExec> * const this) (/home/yhwang/WS_git/presto/presto-native-execution/velox/velox/functions/prestosql/WordStem.h:79)
```

You can see the line numbers in those folly files are different. I am using the folly version from the main branch which is `v2024.04.01.00`. Is it possible the error is from the different folly version?

pedroerp commented:
@yhwang I run some more tests. The problem was because on the previous version you were taking a std::string or StringView's buffer (which do not necessarily have a \0 at the end), and using it in the function sb_stemmer_new() which expects a C-like string ended by a \0.  

Using `c_str()` instead like what you're doing now fixes the issue. I'll get it merged today.

yhwang commented:
@pedroerp thanks for the verification and explanation. So my guess is right :-), although I can’t reproduce the error.
